_posts/2020-11-11--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,222 @@
+---
+layout: post
+title: Mastering AWS Kinesis Data Streams, Part 2
+author: anahitpo
+excerpt: >
+  By now we have gone in quite some details when it comes to writing your dat to a Kinesis stream. How about making use of that data, aka reading it from a stream?
+tags:
+- AWS
+- Data streaming
+- Lambda
+- Serverless
+- Best practices
+---
+
+DO NOT MERGE YET!!! :)
+
+
+This is the second part of the post about AWS _Kinesis Data Streams_. Part 1 is [here](https://dev.solita.fi/2020/05/28/kinesis-streams-part-1.html).
+
+
+# Reading from a Stream
+
+So, you have successfully written your valuable data to the stream, using all the tips from the Part 1 (link). What's next?
+
+## All these consumers
+As with writing to a stream, there are a bunch of ways to read data from a stream. But first, let’s make a short detour.
+
+**TODO**: img of detour
+
+Kinesis family has two more extremely useful services for working with streaming data:
+- _Kinesis Data Firehose_, which is meant to load data to a destination with plenty of them to chose from (think _S3_, _Redshift_, even HTTP endpoints), and
+- _Kinesis Data Analytics_, which can run queries on your streaming data in near real-time. 
+
+I won't be going into details of those services here, each of them deserves a blogpost (or two 😉). I will just say that both, _Kinesis Firehose_ and _Kinesis Data Analytics_ can be used as stream consumers, providing you with out of the box possibilities to analyse the data being streamed, as well as to deliver your data to a destination of your choice. 
+
+One of the main superpowers of _Kinesis Streams_ though is that you can attach **custom data consumers** to it to process and handle data in any way you prefer, in near real-time. There are once again plenty of options to choose from.
+
+If u are inclined to do so, you can use an [EMR cluster as your custom consumer](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-kinesis.html). Or, as with writing to the stream, there are a lot of third-party tools that integrate with _Kinesis Streams_. Think of possibilities like Databricks and Apache Spark integration (which actually uses KCL under the hood), among others.
+
+On a perhaps more familiar side of things, you can have your own consumer application reading from the stream. For that purposes you can once again use either AWS SDK or _Kinesis Consumer Library (KCL)_, which is the counterpart of the _[already familiar KPL](https://dev.solita.fi/2020/05/28/kinesis-streams-part-1.html)_ and which I’ll be mentioning briefly.
```

Comment:
I suppose the paragraphs have been shuffled a bit, so explaning KCL could be moved to the previous paragraph, since the acronym is already used there.